module/zfs/dsl_crypt.c

```diff
@@ -0,0 +1,1585 @@
+	dsl_wrapping_key_t *found_wkey;
+	uint64_t crypt;
+	const char *keysource = skra->skra_cp->cp_keysource;
+
+	/* create and initialize the wrapping key */
+	VERIFY0(dsl_dataset_hold(dp, skra->skra_dsname, FTAG, &ds));
+	ASSERT(!ds->ds_is_snapshot);
+
+	/* set additional properties which can be sent along with this ioctl */
+	if (keysource)
+		dsl_prop_set_sync_impl(ds,
+		    zfs_prop_to_name(ZFS_PROP_KEYSOURCE), ZPROP_SRC_LOCAL,
+		    1, strlen(keysource) + 1, keysource, tx);
+
+	if (skra->skra_cp->cp_iters)
+		dsl_prop_set_sync_impl(ds,
+		    zfs_prop_to_name(ZFS_PROP_PBKDF2_ITERS),
+		    ZPROP_SRC_LOCAL, 8, 1, &skra->skra_cp->cp_iters, tx);
+
+	dsl_prop_set_sync_impl(ds, zfs_prop_to_name(ZFS_PROP_SALT),
+	    ZPROP_SRC_LOCAL, 8, 1, &skra->skra_cp->cp_salt, tx);
+
+	/*
+	 * Rewrapping the dataset implies we are making this an encryption
+	 * root. Set the encryption property locally now.
+	 */
+	VERIFY0(dsl_prop_get_ds(ds, zfs_prop_to_name(ZFS_PROP_ENCRYPTION),
+	    8, 1, &crypt, NULL));
+	dsl_prop_set_sync_impl(ds, zfs_prop_to_name(ZFS_PROP_ENCRYPTION),
+	    ZPROP_SRC_LOCAL, 8, 1, &crypt, tx);
+
+	wkey->wk_ddobj = ds->ds_dir->dd_object;
+
+	rw_enter(&spa->spa_keystore.sk_wkeys_lock, RW_WRITER);
+
+	/* recurse through all children and rewrap their keys */
+	spa_keystore_rewrap_sync_impl(ds->ds_dir->dd_object,
+	    ds->ds_dir->dd_object, wkey, tx);
+
+	/*
+	 * All references to the old wkey should be released now (if it
+	 * existed). Replace the wrapping key.
+	 */
+	found_wkey = avl_find(&spa->spa_keystore.sk_wkeys, wkey, NULL);
+	if (found_wkey) {
+		ASSERT0(refcount_count(&found_wkey->wk_refcnt));
+		avl_remove(&spa->spa_keystore.sk_wkeys, found_wkey);
+		dsl_wrapping_key_free(found_wkey);
+	}
+
+	avl_find(&spa->spa_keystore.sk_wkeys, wkey, &where);
+	avl_insert(&spa->spa_keystore.sk_wkeys, wkey, where);
+
+	rw_exit(&spa->spa_keystore.sk_wkeys_lock);
+
+	dsl_dataset_rele(ds, FTAG);
+}
+
+int
+spa_keystore_rewrap(const char *dsname, dsl_crypto_params_t *dcp)
+{
+	spa_keystore_rewrap_args_t skra;
+
+	/* initialize the args struct */
+	skra.skra_dsname = dsname;
+	skra.skra_cp = dcp;
+
+	/* perform the actual work in syncing context */
+	return (dsl_sync_task(dsname, spa_keystore_rewrap_check,
+	    spa_keystore_rewrap_sync, &skra, 0, ZFS_SPACE_CHECK_NORMAL));
+}
+
+int
+dmu_objset_create_encryption_check(dsl_dir_t *pdd, dsl_crypto_params_t *dcp)
+{
+	int ret;
+	dsl_wrapping_key_t *wkey = NULL;
+	uint64_t cmd = 0, salt = 0, iters = 0;
+	uint64_t pcrypt, crypt = ZIO_CRYPT_INHERIT;
+	const char *keysource = NULL;
+
+	if (!spa_feature_is_enabled(pdd->dd_pool->dp_spa,
+	    SPA_FEATURE_ENCRYPTION) && dcp)
+		return (SET_ERROR(EINVAL));
+
+	ret = dsl_prop_get_dd(pdd, zfs_prop_to_name(ZFS_PROP_ENCRYPTION),
+	    8, 1, &pcrypt, NULL, B_FALSE);
+	if (ret)
+		return (ret);
+
+	if (dcp) {
+		crypt = dcp->cp_crypt;
+		wkey = dcp->cp_wkey;
+		salt = dcp->cp_salt;
+		iters = dcp->cp_iters;
+		keysource = dcp->cp_keysource;
+		cmd = dcp->cp_cmd;
```

Comment:
I could re-implement it using the first method and that would probably be fine. The second one is actually a bit more complicated than it seems because the user can try doing things like `encryption=off` which is completely valid (although unnecessary) in some situations and not others. This would result in the dcp existing with nothing set other than the crypt. I will replace the local variables with a blank struct though as you suggested though.